### DISCUSSION TITLE
Relationship between metadata hashes in explorers vs. Kupo?

### DISCUSSION THREAD

**Discussion by ross-spencer** [CardanoSolutions/kupo#167] [Q&A]
I have a transaction: 

* https://preprod.cardanoscan.io/transaction/455e157fcb9024547922e393b83765606be8a9ddb70299b345f05d6f2f8b883f?tab=metadata

It reports its hash to be: `d5d5742a733382f6101df92a7df5273ce968036ef3e5b25df536b820cd22aba1`

If I access this indexed via Kupo, the hash is reported as: `ae45b0be39e8e84fec8dfddf3c5483c9870e55daad69f0bd8ab092a9d83ccb6b`

E.g.

```json
[
  {
    "hash": "ae45b0be39e8e84fec8dfddf3c5483c9870e55daad69f0bd8ab092a9d83ccb6b",
    "raw": "a11904caa5646461746101676578616d706c6583010203676d65737361676583767468697320697320746865206669727374206c696e65777468697320697320746865207365636f6e64206c696e65767468697320697320746865207468697264206c696e656e736f6d657468696e6720656c7365666161626263636361616163626231",
    "schema": {
      "1226": {
        "map": [
          {
            "k": {
              "string": "data"
            },
            "v": {
              "int": 1
            }
          },
          {
            "k": {
              "string": "example"
            },
            "v": {
              "list": [
                {
                  "int": 1
                },
                {
                  "int": 2
                },
                {
                  "int": 3
                }
              ]
            }
          },
          {
            "k": {
              "string": "message"
            },
            "v": {
              "list": [
                {
                  "string": "this is the first line"
                },
                {
                  "string": "this is the second line"
                },
                {
                  "string": "this is the third line"
                }
              ]
            }
          },
          {
            "k": {
              "string": "something else"
            },
            "v": {
              "string": "aabbcc"
            }
          },
          {
            "k": {
              "string": "aaa"
            },
            "v": {
              "string": "bb1"
            }
          }
        ]
      }
    }
  }
]
```

Is there any reason the hashes should be different in an explorer's view vs. Kupo?

The original data can be used to recreate this with tag: `1226` and:

```json
{
    "data": 1,
    "example": [
        1,
        2,
        3
    ],
    "message": [
        "this is the first line",
        "this is the second line",
        "this is the third line"
    ],
    "something else": "aabbcc",
    "aaa": "bb1"
}
```

Incidentally, the CBOR reported by Kupo maintains the original structure of this snippet perfectly where the explorers currently don't. Still trying to get my head around that too.

**Comment by KtorZ** [comment]
May you indicate what version of Kupo you are currently using? 

There was an issue regarding the calculation of metadata hash that was fixed in `2.7.1` since metadata hash can generally not be safely recomputed from the raw data (since the serialization is non-canonical, one must hold on to the original bytes to ensure the same preimage is used for hashing).  

In this instance, however, it seems even that the raw metadata provided by Kupo is the preimage of the provided hash. Yet I did double-check on another tool which also yields `d5d5742a733382f6101df92a7df5273ce968036ef3e5b25df536b820cd22aba1` as a metadata hash. I will double-check that carefully with a bit more time in front of me to get to the bottom of it.

**Reply by ross-spencer** [reply]
@KtorZ my apologies. I just checked and I am on `v2.5.1` so if that makes a difference here, there probably isn't an issue. (I forgot I had to fall back to an older instance for this node). I'll be on an upgraded node shortly so will try again.
